core/src/subaddress.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Mobilecoin Subaddress Derivations
+
+use curve25519_dalek::{ristretto::RistrettoPoint, scalar::Scalar};
+
+use mc_crypto_hashes::{Blake2b512, Digest};
+use mc_crypto_keys::{RistrettoPrivate, RistrettoPublic};
+
+use crate::{
+    account::{Account, SpendSubaddress},
+    consts::SUBADDRESS_DOMAIN_TAG,
+    keys::*,
+};
+
+/// Generate a subaddress for a given input key set
+pub trait Subaddress {
+    /// Subaddress type
+    type Output: core::fmt::Debug;
+
+    /// Generate the subaddress for the corresponding index
+    fn subaddress(&self, index: u64) -> Self::Output;
+}
+
+/// Generate subadress private keys from root private keys
+impl<'a> Subaddress for (&'a RootViewPrivate, &'a RootSpendPrivate) {
+    type Output = (SubaddrViewPrivate, SubaddrSpendPrivate);
```

Comment:
so this is an outcome of removing duplicated derivation logic... the idea was the tuple-by-reference versions are the One True Impl, then implementations on higher level types can call down without needing to construct a viable higher level object (or copy the keys into this stack frame).

it wouldn't really matter if you -did- need to construct an `Account`, it's just nice if you already have your keys hanging around not to need to create a new object.

Reply:
I could see the by reference argument for HW. Do we have any measurements to know if this provides any performance benefit.

As for higher level types not needing to construct a higher level object.  I would hope if an object had need for a private view and private spend key, the object would have an account, but there are probably some use cases I'm not aware of.

Reply:
on reflection construction is not that big of a deal / it really comes down to how we'd like to express the type relationships. i could just as easily build an `mc_core::Account` or `mc_core::ViewAccount`` to then call `.subaddress()` on... see [here](https://github.com/mobilecoinfoundation/mobilecoin/pull/2525/files#diff-d1e7a41f50ed7d6f5dd04bba9479fa5f52a0b7e2436f73cb5a561b210ea7111fR511) for an example of how i'm currently using the key-based traits, but nbd really to change this.

> I could see the by reference argument for HW. Do we have any measurements to know if this provides any performance benefit.

at the moment i'd estimate we have < 1k free of stack on the ledger (approximately the size of the of a 11 element ring, because we could fit 10 entries but not 11 before reworking the ring signatures for streaming computation). next ledger firmware update should relax that substantially, but i'm perhaps a little over sensitive to memory use this week 😂